File: src/clase-04/ts/factory.ts
// Clase 4 — Patrones Creacionales
// Ejemplo: Factory Method (TypeScript)
// ------------------------------------

// Producto abstracto
interface Shape {
  draw(): string;
}

// Productos concretos
class Circle implements Shape {
  draw(): string {
    return "Dibujando un Círculo";
  }
}

class Square implements Shape {
  draw(): string {
    return "Dibujando un Cuadrado";
  }
}

// Factory
class ShapeFactory {
  createShape(type: string): Shape {
    const t = type.toLowerCase();
    if (t === "circle") return new Circle();
    if (t === "square") return new Square();
    throw new Error(`Forma no reconocida: ${type}`);
  }
}

// Uso
const factory = new ShapeFactory();
const shape1 = factory.createShape("circle");
const shape2 = factory.createShape("square");
console.log(shape1.draw());
console.log(shape2.draw());

File: src/clase-04/python/factory.py
"""
Clase 4 — Patrones Creacionales
Ejemplo: Factory Method (Python)
--------------------------------
Encapsula la creación de objetos. El cliente no usa 'new' directamente.
"""
from abc import ABC, abstractmethod

# Producto abstracto
class Shape(ABC):
    @abstractmethod
    def draw(self) -> str:
        pass

# Productos concretos
class Circle(Shape):
    def draw(self) -> str:
        return "Dibujando un Círculo"

class Square(Shape):
    def draw(self) -> str:
        return "Dibujando un Cuadrado"

# Factory
class ShapeFactory:
    def create_shape(self, shape_type: str) -> Shape:
        st = shape_type.lower()
        if st == "circle":
            return Circle()
        elif st == "square":
            return Square()
        else:
            raise ValueError(f"Forma no reconocida: {shape_type}")

if __name__ == "__main__":
    factory = ShapeFactory()
    shape1 = factory.create_shape("circle")
    shape2 = factory.create_shape("square")
    print(shape1.draw())
    print(shape2.draw())
